crates/shared/src/sources/balancer_v2/pools/stable.rs

```diff
@@ -22,20 +33,164 @@ impl PoolIndexing for PoolInfo {
     }
 }
 
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct PoolState {
+    pub tokens: BTreeMap<H160, common::TokenState>,
+    pub swap_fee: Bfp,
+    pub amplification_parameter: AmplificationParameter,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct AmplificationParameter {
+    factor: U256,
+    precision: U256,
+}
+
+impl AmplificationParameter {
+    pub fn new(factor: U256, precision: U256) -> Result<Self> {
+        ensure!(!precision.is_zero(), "Zero precision not allowed");
+        Ok(Self { factor, precision })
+    }
+
+    /// This is the format used to pass into smart contracts.
+    pub fn as_u256(&self) -> U256 {
+        self.factor * self.precision
+    }
+
+    /// This is the format used to pass along to HTTP solver.
+    pub fn as_big_rational(&self) -> BigRational {
+        // We can assert that the precision is non-zero as we check when constructing
+        // new `AmplificationParameter` instances that this invariant holds, and we don't
+        // allow modifications of `self.precision` such that it could become 0.
+        debug_assert!(!self.precision.is_zero());
+        BigRational::new(self.factor.to_big_int(), self.precision.to_big_int())
+    }
+}
+
 #[async_trait::async_trait]
 impl FactoryIndexing for BalancerV2StablePoolFactory {
     type PoolInfo = PoolInfo;
+    type PoolState = PoolState;
 
     async fn specialize_pool_info(&self, pool: common::PoolInfo) -> Result<Self::PoolInfo> {
         Ok(PoolInfo { common: pool })
     }
+
+    fn fetch_pool_state(
```

Comment:
`get_pool_tokens` returns the reserves. This was the same as the old code:

https://github.com/gnosis/gp-v2-services/blob/8189f04dba1c8f9a88a2191374799649b26fcf85/crates/shared/src/sources/balancer_v2/pool_cache.rs#L88-L92

Reply:
Maybe we could rename this to get pool reserves which implies the tokens must code along with it.

Reply:
note that `getPoolTokens` is the name of the contract method we are calling on the Vault here. My worry about renaming it is that it will cause confusion where someone sees we are calling `get_reserves` and then go to the `IVault.sol` file and don't see that method implemented anywhere.